flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/util/ExplodeFunctionUtil.scala

```diff
@@ -18,74 +18,244 @@
 
 package org.apache.flink.table.plan.util
 
+import java.util
+
 import org.apache.flink.api.common.typeinfo.{BasicArrayTypeInfo, BasicTypeInfo, PrimitiveArrayTypeInfo, TypeInformation}
-import org.apache.flink.api.java.typeutils.ObjectArrayTypeInfo
+import org.apache.flink.api.java.typeutils.{MultisetTypeInfo, ObjectArrayTypeInfo}
 import org.apache.flink.table.functions.TableFunction
 
 class ObjectExplodeTableFunc extends TableFunction[Object] {
   def eval(arr: Array[Object]): Unit = {
-    arr.foreach(collect)
+    if (null != arr) {
+      var i = 0
```

Comment:
Can't you just do the following?

` if (null != arr) {
     arr.foreach(collect)
}`

Reply:
As Timo mentioned above "We should not use any Scala magic in runtime code." So I convert the foreach loops in all the methods of this class."

Reply:
Make sense. The JVM might not be smart enough to inline foreach().